### DISCUSSION TITLE
SoapClient `3.3.0`: Any hints how to generate class with default value initialization?

### DISCUSSION THREAD

**Discussion by trailsnail** [phpro/soap-client#529] [Q&A] [answered]
Are there some best practices on how to generate soap classes with assigning default values?

I try to explain the problem with following class as example:

```php
class ShipmentMethodTranslations
{
    /**
     * @var array<int<0,max>, \WebConnector\MasterData\Type\ShipmentMethodTranslation>
     */
    private array $Shipment_Method_Translation;

    /**
     * @return array<int<0,max>, \WebConnector\MasterData\Type\ShipmentMethodTranslation>
     */
    public function getShipmentMethodTranslation() : array
    {
        return $this->Shipment_Method_Translation;
    }

    /**
     * @param array<int<0,max>, \WebConnector\MasterData\Type\ShipmentMethodTranslation> $Shipment_Method_Translation
     * @return static
     */
    public function withShipmentMethodTranslation(array $Shipment_Method_Translation) : static
    {
        $new = clone $this;
        $new->Shipment_Method_Translation = $Shipment_Method_Translation;

        return $new;
    }
}

```

We receive ShipmentMethods from API. These ShipmentMethod objects contain a property `ShipmentMethodTranslations`.
But not every translation is filled on API side.
What we do is to call:

`ShipmentMethodTranslations::getShipmentMethodTranslation(): array`

But `ShipmentMethodTranslations`'s `private array $Shipment_Method_Translation;` isn't initialized, due to not existing data on API side.
This results in following error:

```
Typed property WebConnector\MasterData\Type\ShipmentMethodTranslations::$Shipment_Method_Translation must not be accessed before initialization
```

In my opinion it makes sense to generate the class with property with a default value:

```php
private array $Shipment_Method_Translation = [] // define default direct at property;
```

Maybe a `__construct()` would also work, but could be more complicated due to knowledge whar as to be injected.

Any thoughts on this?

Wish a nice day 🌞

**Comment by trailsnail** [comment]
Did some debugging and found while generating classes, that the default value `null` seems to be omited.
But property is defined also not nullalbe with `nullable: false`

```
Laminas\Code\Generator\ClassGenerator {#2858
  #isSourceDirty: true
  #indentation: "    "
  #sourceContent: null
  #containingFileGenerator: null
  #namespaceName: "WebConnector\MasterData\Type"
  #docBlock: null
  #name: "ShipmentMethodTranslations"
  #flags: 0
  #extendedClass: null
  #implementedInterfaces: []
  #properties: array:1 [
    "Shipment_Method_Translation" => Laminas\Code\Generator\PropertyGenerator {#2798
      #isSourceDirty: true
      #indentation: "    "
      #sourceContent: null
      #docBlock: Laminas\Code\Generator\DocBlockGenerator {#2797
        #isSourceDirty: true
        #indentation: ""
        #sourceContent: null
        #shortDescription: ""
        #longDescription: ""
        #tags: array:1 [
          0 => Laminas\Code\Generator\DocBlock\Tag {#2796
            #isSourceDirty: true
            #indentation: "    "
            #sourceContent: null
            #name: "var"
            #content: "array<int<0,max>, \WebConnector\MasterData\Type\ShipmentMethodTranslation>"
          }
        ]
        #wordwrap: false
      }
      #name: "Shipment_Method_Translation"
      #flags: 64
      #isConst: false
      #type: Laminas\Code\Generator\TypeGenerator {#2795
        -type: Laminas\Code\Generator\TypeGenerator\AtomicType {#2793
          +sortIndex: 5
          +type: "array"
        }
        -nullable: false // <=== not nullable
      }
      #defaultValue: null
      -omitDefaultValue: true // <=== no default value
    }
  ]

```

**Reply by veewee** [reply]
Nullable means something else than an empty array though.

You can change how the code is generated by using the https://github.com/phpro/soap-client/blob/v4.x/src/Phpro/SoapClient/CodeGenerator/Assembler/PropertyAssemblerOptions.php.

https://github.com/phpro/soap-client/blob/59a5c10208941b048d52d4786459a4408e6b6cc3/src/Phpro/SoapClient/CodeGenerator/Assembler/PropertyAssemblerOptions.php#L69-L74


I'll have to check how this behaves in the new encoding system (v4) though.
(checked : should work as expected on v4 : https://github.com/php-soap/encoding/blob/f2f7de28089dbdad454c1c4f9337c47e0d857619/src/Encoder/ObjectEncoder.php#L143 )

**Reply by trailsnail** [reply]
Found a solution to generate default `[]` onto array properties - thanks.

But while investigation the question came up, what is SoapClients behavior if `mixed="true"` enters the game.

Given following wsdl:

```
<xsd:complexType name="Shipment_Method">
<xsd:sequence>
<xsd:element minOccurs="0" maxOccurs="1" name="Code" type="xsd:string"/>
<xsd:element minOccurs="0" maxOccurs="1" name="Description" type="xsd:string"/>
<xsd:element minOccurs="0" maxOccurs="1" name="Shipment_Method_Translations" type="q26:Shipment_Method_Translations"/>
</xsd:sequence>
</xsd:complexType>
```

Results in following correct class properties:

```php
class ShipmentMethod
{
    /**
     * @var null | string
     */
    private ?string $Code = null;

    /**
     * @var null | string
     */
    private ?string $Description = null;

	...
```

But then we have this wsdl:
```
<xsd:complexType name="Shipment_Method_Translations" mixed="true">
<xsd:sequence>
<xsd:element minOccurs="0" maxOccurs="unbounded" name="Shipment_Method_Translation" type="q27:Shipment_Method_Translation"/>
</xsd:sequence>
</xsd:complexType>
```

What does this `mixed="true"` do, since the result is a class property `$Shipment_Method_Translation` without usefull default.

```php
class ShipmentMethodTranslations
{
    /**
     * @var array<int<0,max>, \WebConnector\MasterData\Type\ShipmentMethodTranslation>
     */
    private array $Shipment_Method_Translation;
	...

```

This question came up, because API Response does not contain `$Shipment_Method_Translation` (optional) - and so trhe property was never initialized 👈🏻 Error on first access in code

Maybe if there is a problem while interpreting WSDL, the the solution should be somewhere else instead of adding empty arrays with Assembler

**Reply by veewee** [reply]
I don't think the wsdl-reader is parsing that `mixed` info at the moment, so it should fall back to the array default of `[]`.
It's also a very confusing property that allows a mix of text and elements.
I'm not sure how it should behave. An example can be found here:

https://www.w3schools.com/xml/schema_complex_mixed.asp

```xml
<letter>
  Dear Mr. <name>John Smith</name>.
  Your order <orderid>1032</orderid>
  will be shipped on <shipdate>2001-07-13</shipdate>.
</letter> 
```

```xml

<xs:element name="letter">
  <xs:complexType mixed="true">
    <xs:sequence>
      <xs:element name="name" type="xs:string"/>
      <xs:element name="orderid" type="xs:positiveInteger"/>
      <xs:element name="shipdate" type="xs:date"/>
    </xs:sequence>
  </xs:complexType>
</xs:element> 
```

Feel free to test it out on regular ext-soap to see how it works with mixed content and enlightn me :)

**Reply by trailsnail** [reply]
`mixed="true"` has nothing to do with our problem - we found out. 😉 Hope you are enlighted now 🤣 

But we found this while debugging more: https://github.com/phpro/soap-client/issues/515#issuecomment-2114914477 which describes our problem we have with arrays.
Seems that there is missing some default behavior for arrays which are defined as `minOccurs="0" maxOccurs="unbounded"` without a `nillable="true"` attribute.

The question is: should they be `null` by default too? Or should I use our new `DefaultValueForEmptyArrayAssembler` 🤣 



to be more specific:

```
<xsd:complexType name="Shipment_Method_Translations" mixed="true">
	<xsd:sequence>
	<xsd:element minOccurs="0" maxOccurs="unbounded" name="Shipment_Method_Translation" type="q27:Shipment_Method_Translation"/>
	</xsd:sequence>
</xsd:complexType>
```

should result in:

```php
 /**
     * @var array<int<0,max>, \WebConnector\MasterData\Type\ShipmentMethodTranslation>
     */
    private array $Shipment_Method_Translation = []
```

and 

```
<xsd:complexType name="Shipment_Method_Translations" mixed="true">
	<xsd:sequence>
	<xsd:element minOccurs="0" maxOccurs="unbounded" nillable="true" name="Shipment_Method_Translation" type="q27:Shipment_Method_Translation"/>
	</xsd:sequence>
</xsd:complexType>
```

should result in:

```php
 /**
     * @var null | array<int<0,max>, \WebConnector\MasterData\Type\ShipmentMethodTranslation>
     */
    private ?array $Shipment_Method_Translation = null
```

Last example width `nillable` WSDL already works.

**Reply by veewee** [reply, marked as answer]
Those rules are being covered by the code generator though:

See https://github.com/phpro/soap-client/blob/v4.x/src/Phpro/SoapClient/CodeGenerator/TypeEnhancer/MetaTypeEnhancer.php


Our assemblers are not the once setting default values.
That is what laminas-code is doing for you. If you want to overwrite the default value - you should be configuring this through a custom property assembler:

https://github.com/laminas/laminas-code/blob/562e02b7d85cb9142b5116cc76c4c7c162a11a1c/src/Generator/PropertyGenerator.php#L42-L44

If done right, it could be added to this package maybe.

However, the problem you are really facing, is the fact that ext-soap is skipping setting the property during hydration instead of `[]`. This is what our new custom encoder (for v4 which is in alpha - [you should definetly check it out](https://github.com/phpro/soap-client/releases/tag/4.0.0-alpha1) ;-) ) is doing right. Therefore, you won't get that uninitialized property error anymore and you wouldn't care about the default of the property anymore.

**Reply by trailsnail** [reply]
Then we found a solution: set a default value `[]` for array properties - already working, were some hard last days migrate from SoapClient `1.7.4` to `3.3.0`: but now configuration feels more easy - Good work 👍🏻
